### PR TITLE
Ensure that tidy follows the same procedure for sorting

### LINKED ISSUE #8734: Ensure that tidy follows the same procedure for sorting
#8692 introduced a new way to sort the `crate` declarations (thanks to @GuillaumeGomez). The current procedure for sorting `mod` and `use` declarations [makes use of the `sorted` function](https://github.com/servo/servo/blob/6e7de62b38a04e614c60e864c8d100bf3e72304f/python/tidy.py#L345). This means that every time tidy encounters a declaration, it's appended to the collected list of declarations and sorts them again. I think that's inefficient (and may consume time, given the abundance of files in Servo). So, let's modify both the procedures to stick to the one followed for crate declarations :)

Code: `python/tidy.py`

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #8734 

cc @Wafflespeanut

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8750)

### CODE REVIEW

**Comment by jdm** [comment]
@Wafflespeanut Want to review this one?

**Comment by Wafflespeanut** [comment]
@jdm Please don't ask! Just delegate me! :P

---

Reviewed 1 of 1 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[python/tidy.py, line 223 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RAOLbMW7LbryGgapY:-K4RAOLbMW7LbryGgapZ:512831837) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L223)):*
I'm curious. Why aren't we going for a dictionary here? `use` declarations could've been indented, right?

---

*[python/tidy.py, line 242 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RCuub0FMpo6GM9uyB:-K4RCuud4kAkWhFXxizX:1683290886) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L242)):*
Let's move these three strings to the top (say `decl_msg`, `decl_expect`, `decl_found`), so that whenever we need one, we can just format it on the fly.

---

*[python/tidy.py, line 343 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RAvZZdtfIgfkrUb3f:-K4RAvZZdtfIgfkrUb3g:486415127) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L343)):*
This line's repeated for all the cases. So, let's move it to the start, and more importantly, let's group all the cases together. Now, `extern crate` check is at the top, while the other two are at the bottom.

---

*[python/tidy.py, line 369 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RDo3IeXNQvRzJIfr0:-K4RDo3JG9OGoGe-J7W9:893467459) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L369)):*
Not sure why we'll be needing this funny thing (after all, it's just getting the name, right?). This could be done in a single line.

---

*[python/tidy.py, line 376 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RH7ZOutmaOZX25fq3:-K4RH7ZQfzbAPSqn8RNy:2111080895) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L376)):*
... could be joined with `and`

---

*[python/tidy.py, line 385 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RHVVcgTudXmquW8QT:-K4RHVVdZtnmDYhvPNaM:216880679) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L385)):*
I'm not sure why we need a `prev_mod_indent` - it's only been used in two places, and it's value's never changed from 0.

---

*[python/tidy.py, line 392 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RJQnzVH6-oshwfU9J:-K4RJQo-C9EWXNSgWfyL:-1818282155) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L392)):*
If we're just erasing the entries, then we could go for `prev_mod.clear()`, no?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750#-:-K4RAZlb2g_K06Hqb9gD:-1793851538)*

**Comment by Wafflespeanut** [comment]
Reviewed 1 of 1 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions.

---

*[components/compositing/scrolling.rs, line 10 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RLRwMPXAf--f08N0W:-K4RLRwMPXAf--f08N0X:-1421904704) ([raw file](https://github.com/servo/servo/blob/0f7204936394fe104431f01c68ffa17829426cba/components/compositing/scrolling.rs#L10)):*
This should probably be another PR, because our cleanup's gonna take some time! (when this would already be removed) :P

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750)*

**Comment by GuillaumeGomez** [comment]
*[components/compositing/scrolling.rs, line 10 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RLRwMPXAf--f08N0W:-K4Rt1UnUplE_7WmWAcY:1757376908) ([raw file](https://github.com/servo/servo/blob/0f7204936394fe104431f01c68ffa17829426cba/components/compositing/scrolling.rs#L10)):*
I didn't even pay attention ! :O

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750)*

**Comment by GuillaumeGomez** [comment]
*[python/tidy.py, line 223 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RAOLbMW7LbryGgapY:-K4RtAE1X_d-KtcfI0aB:-1143085781) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L223)):*
They could ? In which way ? You mean :

```Rust
use something::other::{
    1,
    2,
};
```
?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750)*

**Comment by GuillaumeGomez** [comment]
*[python/tidy.py, line 369 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RDo3IeXNQvRzJIfr0:-K4RtaOjB8J3sEiwSMOU:2069966223) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L369)):*
How would you do it in a single line ?

```Python
mod = line[(line.startswith("mod ") ? 4 : 8):]
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750)*

**Comment by GuillaumeGomez** [comment]
*[python/tidy.py, line 376 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RH7ZOutmaOZX25fq3:-K4RuLIr7ndAZE_zZ60t:241276095) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L376)):*
Could be joined with what ?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750)*

**Comment by GuillaumeGomez** [comment]
*[python/tidy.py, line 392 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RJQnzVH6-oshwfU9J:-K4RuUCGL3tr5aexgLor:1211507093) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L392)):*
I thought about it for a while, but if I call clear, I'll have to create again key entries. So either we set back the key value to "", or we have to recreate all key entries.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750)*

**Comment by GuillaumeGomez** [comment]
*[python/tidy.py, line 385 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RHVVcgTudXmquW8QT:-K4RvibT_u4Nuohgsz8f:-247868917) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L385)):*
I forgot to remove it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750)*

**Comment by Wafflespeanut** [comment]
It looks like some of my comments haven't been addressed yet...

---

Reviewed 1 of 1 files at r3, 3 of 3 files at r4.
Review status: 1 of 2 files reviewed at latest revision, 7 unresolved discussions.

---

*[python/tidy.py, line 223 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RAOLbMW7LbryGgapY:-K4SJtLOasZtCrRTy33C:-1248123106) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L223)):*
I meant something like, 

``` rust
use foo::bar;
#[test]
fn foobar {
    use bar::baz;
}
```

Of course, we have cases like these in unit tests.

---

*[python/tidy.py, line 242 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RCuub0FMpo6GM9uyB:-K4S3CJyzYH0fawupoQw:824973325) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L242)):*
[not addressed]

---

*[python/tidy.py, line 343 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RAvZZdtfIgfkrUb3f:-K4S3H6YblCvF6k-xZsM:824973325) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L343)):*
[not addressed]

---

*[python/tidy.py, line 369 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RDo3IeXNQvRzJIfr0:-K4S-UE-KgcxLkgCiJEk:-1819864734) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L369)):*
``` python
mod = line[4:] if line.startswith("mod ") else line[8:]
```

No need for `mod = ""` if we go for the "Pythonic" way :)

---

*[python/tidy.py, line 376 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RH7ZOutmaOZX25fq3:-K4S-sM2et_6qFLPAMi8:-1324584047) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L376)):*
```python
if match == -1 and not mod.endswith(";"):
```

... since we're interested in only one case here

---

*[python/tidy.py, line 392 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RJQnzVH6-oshwfU9J:-K4S2DbpQLGuZy4gz4ps:1825124480) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L392)):*
I dunno, (given the number of `mod` declarations in a file) I doubt this would make any difference :)

---

*[python/tidy.py, line 326 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4SWuBsKMQGiLCOEaX1:-K4SWuBtM72AmGNNCT-i:1167991508) ([raw file](https://github.com/servo/servo/blob/2ef68a81d14093b74eeeffca7bb9cc4111cd94e8/python/tidy.py#L326)):*
Oh, and we know that the last character is semi-colon, right? So, let's use slicing, like `[:-1]`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750#-:-K4S-NZCJ_FnVrByapEj:-1433781864)*

**Comment by GuillaumeGomez** [comment]
*[python/tidy.py, line 223 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RAOLbMW7LbryGgapY:-K4SaxSpy23A_VCIM6qC:942185798) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L223)):*
Oh ! I'm ashamed to haven't thought about it. :(

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750)*

**Comment by GuillaumeGomez** [comment]
*[python/tidy.py, line 343 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RAvZZdtfIgfkrUb3f:-K4Rs_DmMzOJRW_MM8cz:92950814) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L343)):*
This line is repeated to avoid calling when it's not necessary. However, I should indeed move "extern crate" check to the bottom.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750)*

**Comment by GuillaumeGomez** [comment]
*[python/tidy.py, line 223 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RAOLbMW7LbryGgapY:-K4UBsSgQmrLiyreQYUE:-727600147) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L223)):*
I actually already handled this case, but differently. When indentation changes, I stop checking for previous indent use statement (in some case, the order is very important). So if I find a `#[macro_use]`, I restart checking the order too. I hope I made my explanations understandable. :-/

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750)*

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by Wafflespeanut** [comment]
Reviewed 1 of 1 files at r3, 3 of 3 files at r4, 1 of 1 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[python/tidy.py, line 228 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4WzaGZM4cGpiZI9yoW:-K4WzaGZM4cGpiZI9yoX:1050634656) ([raw file](https://github.com/servo/servo/blob/b938046dfd0f06855a3fda9f4c3edd2f0b8f0012/python/tidy.py#L228)):*
Nit: This should actually be `"{} is not in alphabetical order" - we'll just stick to what it was previously - `use` "statements", `mod` & `extern crate` "declarations".

---

*[python/tidy.py, line 330 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4X-ti58mOXxS0wxn2j:-K4X-ti58mOXxS0wxn2k:1412370650) ([raw file](https://github.com/servo/servo/blob/b938046dfd0f06855a3fda9f4c3edd2f0b8f0012/python/tidy.py#L330)):*
Sorry for changing this again (I must've overlooked this at first). Since we've already checked whether the line starts with `extern crate`, let's put this like `[13:-1]` (12 or 13? check the off-by-one error, btw :P)

---

*[python/tidy.py, line 382 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4X1HH5Smbe6P26kPU4:-K4X1HH5Smbe6P26kPU5:329637312) ([raw file](https://github.com/servo/servo/blob/b938046dfd0f06855a3fda9f4c3edd2f0b8f0012/python/tidy.py#L382)):*
I'm not sure if we should really erase this. What if we just leave this `else`? Can you check whether this works as expected? (i.e., doesn't break tidy)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750)*

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by SimonSapin** [comment]
Reviewed 1 of 1 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[python/tidy.py, line 330 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4X-ti58mOXxS0wxn2j:-K4cloTJl3ro8bvun9dY:53378151) ([raw file](https://github.com/servo/servo/blob/b938046dfd0f06855a3fda9f4c3edd2f0b8f0012/python/tidy.py#L330)):*
Sorry contradict @waffespeanut, but I’d rather the `13` magic number be `len("extern crate ")` instead. (And the space after crate be added in `startswith` as well.)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750)*

**Comment by GuillaumeGomez** [comment]
Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions.

---

*[python/tidy.py, line 369 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RDo3IeXNQvRzJIfr0:-K4dbyqeNmJ7or1ECM7_:-499146009) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L369)):*
Done.

---

*[python/tidy.py, line 376 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RH7ZOutmaOZX25fq3:-K4dc-Edom2AKvTTpws3:-499146009) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L376)):*
Done.

---

*[python/tidy.py, line 392 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4RJQnzVH6-oshwfU9J:-K4dc0aIPQD43RfwaAwU:-499146009) ([raw file](https://github.com/servo/servo/blob/14eaac29c5bfba9ebe3f4153e5a990148df1bb64/python/tidy.py#L392)):*
Done.

---

*[python/tidy.py, line 228 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4WzaGZM4cGpiZI9yoW:-K4dbvFuN5WPBr9h4QZ0:-499146009) ([raw file](https://github.com/servo/servo/blob/b938046dfd0f06855a3fda9f4c3edd2f0b8f0012/python/tidy.py#L228)):*
Done.

---

*[python/tidy.py, line 330 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4X-ti58mOXxS0wxn2j:-K4dbxpmTWkxFw-YU2XJ:-499146009) ([raw file](https://github.com/servo/servo/blob/b938046dfd0f06855a3fda9f4c3edd2f0b8f0012/python/tidy.py#L330)):*
Done.

---

*[python/tidy.py, line 382 \[r6\]](https://reviewable.io:443/reviews/servo/servo/8750#-K4X1HH5Smbe6P26kPU4:-K4dc3b0WR--dsgfXmhl:-499146009) ([raw file](https://github.com/servo/servo/blob/b938046dfd0f06855a3fda9f4c3edd2f0b8f0012/python/tidy.py#L382)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8750)*

**Comment by Ms2ger** [comment]
This PR seems to accidentally revert the WPT link check.

**Comment by GuillaumeGomez** [comment]
Oh? I thought I revert the accidentaly reverted WPT link check...

**Comment by GuillaumeGomez** [comment]
Updated.

File: python/tidy.py [hunk 1]
```diff
@@ -366,26 +371,24 @@ def check_rust(file_name, contents):
...
+                    yield (idx + 1, "mod declaration spans multiple lines")
+                mod = mod[:len(mod) - 1]
+                if len(prev_mod[indent]) > 0 and mod < prev_mod[indent]:
+                    yield(idx + 1, decl_message.format("mod declaration")
+                          + decl_expected.format(prev_mod[indent])
+                          + decl_found.format(mod))
+                prev_mod[indent] = mod
+        else:
+            # we now erase previous entries
+            prev_mod = {}
```
Comment (Wafflespeanut) [inline python/tidy.py]: shouldn't `prev_crate` be also erased at this point?

**Comment by Wafflespeanut** [comment]
Once that's clarified, let's land this! :)

File: python/tidy.py [hunk 2]
```diff
@@ -342,23 +334,36 @@ def check_rust(file_name, contents):
...
+        if line.startswith("extern crate "):
+            crate_name = line[len("extern crate "):-1]
+            indent = len(original_line) - len(line)
+            if indent not in prev_crate:
+                prev_crate[indent] = ""
+            if prev_crate[indent] > crate_name:
+                yield(idx + 1, decl_message.format("extern crate declaration")
+                      + decl_expected.format(prev_crate[indent])
+                      + decl_found.format(crate_name))
+            prev_crate[indent] = crate_name
```
Comment (Wafflespeanut) [inline python/tidy.py]: shouldn't `prev_crate` be erased if the line doesn't start with `extern crate`?

Comment (GuillaumeGomez) [inline python/tidy.py]: No, it's very differerent from `use`. There is no limit on the order, on the contrary of `use` imports. So the sorting is for the whole file.

**Comment by GuillaumeGomez** [comment]
About the squash, I thought splitting out changing `tidy` algorithms in its own commit was a good thing. Merging the mods sorting and these algorithm changes would be a bit too much in my opinion. But if you insist ?

**Comment by Wafflespeanut** [comment]
@GuillaumeGomez I don't see how both the commits ([1](https://github.com/GuillaumeGomez/servo/commit/55676848cbe09d0da2b4f09d1448e9561be2687e), [2](https://github.com/GuillaumeGomez/servo/commit/8c2257f887abed054cd64bfcbef397f0b06cdec2)) are independently helpful? One seems to remove the wpt-check, whereas the other seems to revert it.

**Comment by GuillaumeGomez** [comment]
This wpt-check issue is weird. Hold on, I merge them both.

**Comment by GuillaumeGomez** [comment]
Merged they are.

**Comment by Wafflespeanut** [comment]
@bors-servo r+

Thanks for being patient! :)

**Comment by GuillaumeGomez** [comment]
Thanks for your review ! :)

**Pull request merged**